src/unitquaternion.jl

```diff
@@ -0,0 +1,428 @@
+
+        2 * (yz + xw)
+    elseif i == 7
+        xz = (q.x * q.z)
+        yw = (q.y * q.w)
+
+        2 * (xz + yw)
+    elseif i == 8
+        yz = (q.y * q.z)
+        xw = (q.w * q.x)
+
+        2 * (yz - xw)
+    elseif i == 9
+        ww = (q.w * q.w)
+        xx = (q.x * q.x)
+        yy = (q.y * q.y)
+        zz = (q.z * q.z)
+
+        ww - xx - yy + zz
+    else
+        throw(BoundsError(r,i))
+    end
+end
+
+function Base.Tuple(q::UnitQuaternion)
+    ww = (q.w * q.w)
+    xx = (q.x * q.x)
+    yy = (q.y * q.y)
+    zz = (q.z * q.z)
+    xy = (q.x * q.y)
+    zw = (q.w * q.z)
+    xz = (q.x * q.z)
+    yw = (q.y * q.w)
+    yz = (q.y * q.z)
+    xw = (q.w * q.x)
+
+    # initialize rotation part
+    return (ww + xx - yy - zz,
+            2 * (xy + zw),
+            2 * (xz - yw),
+            2 * (xy - zw),
+            ww - xx + yy - zz,
+            2 * (yz + xw),
+            2 * (xz + yw),
+            2 * (yz - xw),
+            ww - xx - yy + zz)
+end
+
+# ~~~~~~~~~~~~~~~ Getters ~~~~~~~~~~~~~~~ #
+map_type(::UnitQuaternion{T,D}) where {T,D} = D
+map_type(::Type{UnitQuaternion{T,D}}) where {T,D} = D
+
+scalar(q::UnitQuaternion) = q.w
+vector(q::UnitQuaternion{T}) where T = SVector{3,T}(q.x, q.y, q.z)
+
+SVector(q::UnitQuaternion{T}) where T = SVector{4,T}(q.w, q.x, q.y, q.z)
+
+# ~~~~~~~~~~~~~~~ Initializers ~~~~~~~~~~~~~~~ #
+Base.rand(::Type{<:UnitQuaternion{T,D}}) where {T,D} =
+    normalize(UnitQuaternion{T,D}(randn(T), randn(T), randn(T), randn(T)))
+Base.rand(::Type{UnitQuaternion{T}}) where T = Base.rand(UnitQuaternion{T,DEFAULT_QMAP})
+Base.rand(::Type{UnitQuaternion}) = Base.rand(UnitQuaternion{Float64,DEFAULT_QMAP})
+Base.zero(::Type{Q}) where Q<:UnitQuaternion = Q(I)
+Base.zero(q::Q) where Q<:UnitQuaternion = Q(I)
+@inline Base.one(::Type{Q}) where Q <: UnitQuaternion = Q(1.0, 0.0, 0.0, 0.0)
+
+
+# ~~~~~~~~~~~~~~~ Math Operations ~~~~~~~~~~~~~~~ #
+
+# Inverses
+conj(q::UnitQuaternion{T,D}) where {T,D} = UnitQuaternion{T,D}(q.w, -q.x, -q.y, -q.z)
+inv(q::UnitQuaternion) = conj(q)
+(-)(q::UnitQuaternion{T,D}) where {T,D} = UnitQuaternion{T,D}(-q.w, -q.x, -q.y, -q.z)
+
+# Norms
+LinearAlgebra.norm(q::UnitQuaternion) = sqrt(q.w^2 + q.x^2 + q.y^2 + q.z^2)
+vecnorm(q::UnitQuaternion) = sqrt(q.x^2 + q.y^2 + q.z^2)
+
+function LinearAlgebra.normalize(q::UnitQuaternion{T,D}) where {T,D}
+    n = 1/norm(q)
+    UnitQuaternion{T,D}(q.w*n, q.x*n, q.y*n, q.z*n)
+end
+
+# Identity
+(::Type{Q})(I::UniformScaling) where Q <: UnitQuaternion = one(Q)
+
+# # Equality
+# (≈)(q::UnitQuaternion, u::UnitQuaternion) = q.w ≈ u.w && q.x ≈ u.x && q.y ≈ u.y && q.z ≈ u.z
+# (==)(q::UnitQuaternion, u::UnitQuaternion) = q.w == u.w && q.x == u.x && q.y == u.y && q.z == u.z
+
+# Exponentials and Logarithms
+function exp(q::UnitQuaternion{T,D}) where {T,D}
+    θ = vecnorm(q)
+    sθ,cθ = sincos(θ)
+    es = exp(q.w)
+    M = es*sθ/θ
+    UnitQuaternion{T,D}(es*cθ, q.x*M, q.y*M, q.z*M, false)
+end
+
+function expm(ϕ::SVector{3,T}) where T
```

Comment:
This is great feedback. I tend to over-use the type parameters. I've gone back through and eliminated a lot of the unnecessary parameters. 

So in this particular case, if the user passes in an `SVector{3,Int}` then the function will throw an `InexactError`. Is it considered better programming practice in Julia to allows `Int`s only to have them error out (which is an arguably more informative error than "method not found") or to simply not define the method for `eltype(Int)`?

Reply:
The best practice would be to make things "just work" by promoting the incoming `T` to a wider numeric type. As a very simple example, `1/2` doesn't throw an error, because `/` promotes those integers to `Float64` before doing a floating point division.

It's best not to think about explicitly computing the output eltype from `T` if you can avoid it; instead you can do the computation of the elements and use whatever type they happen to have as the output eltype. This leaves the type computations up to the compiler which makes the code neater, and the compiler is very good at figuring these kinds of things out.

So in this particular case, you're going to do some operations such as `norm`, `sincos`, `sinc` which will already have promotion built in, so the type of the elements `cθ, ϕ[1]*M, ϕ[2]*M, ϕ[3]*M` will no longer be `T`.  The you can just pass these to the constructor and use `promote` there (rather than `promote_type`).

To summarize: always strive to compute in the value domain, not the type domain. Explicitly computing types is difficult to do correctly in generic code and is rarely necessary.